server/src/pages/shot/editor.js

```diff
@@ -34,23 +51,278 @@ exports.Editor = class Editor extends React.Component {
+          return {box: 0, pos: 'tr'};
+        } else if (box.y2 - lineOffset < y && y < box.y2 + lineOffset) {
+          return {box: 0, pos: 'br'};
+        } else if (yCenter - lineOffset < y && y < yCenter + lineOffset) {
+          return {box: 0, pos: 'r'};
+        }
+      } else if (xCenter - lineOffset < x && x < xCenter + lineOffset) {
+        if (box.y1 - lineOffset < y && y < box.y1 + lineOffset) {
+          return {box: 0, pos: 't'};
+        } else if (box.y2 - lineOffset < y && y < box.y2 + lineOffset) {
+          return {box: 0, pos: 'b'};
+        } else if (box.y1 - lineOffset < y && y < box.y2 + lineOffset) {
+          return {box: 0, pos: 'i'};
+        }
+      } else if (box.x1 - lineOffset < x && x < box.x2 + lineOffset) {
+        if (box.y1 - lineOffset < y && y < box.y2 + lineOffset) {
+          return {box: 0, pos: 'i'};
+        }
+      }
+    }
+    return {box: -1, pos: 'o'};
+  }
+
+  mousemove(e) {
+    if (mousedown && this.clickedArea.box == -1 && boxes.length < 1) {
+      x2 = e.offsetX;
+      y2 = e.offsetY;
+      this.redraw();
+    } else if (mousedown && this.clickedArea.box != -1) {
+      x2 = e.offsetX;
+      y2 = e.offsetY;
+      let xOffset = x2 - x1;
+      let yOffset = y2 - y1;
+      x1 = x2;
+      y1 = y2;
+      if (this.clickedArea.pos == 'i' ||
+        this.clickedArea.pos == 'tl' ||
+        this.clickedArea.pos == 'l' ||
+        this.clickedArea.pos == 'bl') {
+        boxes[this.clickedArea.box].x1 += xOffset;
+      }
+      if (this.clickedArea.pos == 'i' ||
+        this.clickedArea.pos == 'tl' ||
+        this.clickedArea.pos == 't' ||
+        this.clickedArea.pos == 'tr') {
+        boxes[this.clickedArea.box].y1 += yOffset;
+      }
+      if (this.clickedArea.pos == 'i' ||
+        this.clickedArea.pos == 'tr' ||
+        this.clickedArea.pos == 'r' ||
+        this.clickedArea.pos == 'br') {
+        boxes[this.clickedArea.box].x2 += xOffset;
+      }
+      if (this.clickedArea.pos == 'i' ||
+        this.clickedArea.pos == 'bl' ||
+        this.clickedArea.pos == 'b' ||
+        this.clickedArea.pos == 'br') {
+        boxes[this.clickedArea.box].y2 += yOffset;
+      }
+      this.redraw();
+    }
+  }
+
+  redraw() {
+    let boxX1 = x1 < x2 ? x1 : x2;
+    let boxY1 = y1 < y2 ? y1 : y2;
+    let boxX2 = x1 > x2 ? x1 : x2;
+    let boxY2 = y1 > y2 ? y1 : y2;
+    let context = this.cropper.getContext('2d');
+    context.clearRect(0, 0, this.canvasWidth, this.canvasHeight);
+    context.beginPath();
+    if (boxes.length == 1) {
+      this.drawCropBox(boxes[0], context);
+    }
+    if (this.clickedArea.box == -1) {
+      if (boxX2 - boxX1 > lineOffset * 2 && boxY2 - boxY1 > lineOffset * 2) {
+        this.tempBox = {x1: boxX1, y1: boxY1, x2: boxX2, y2: boxY2}
+      }
+      if (this.tempBox) {
+        this.drawCropBox(this.tempBox, context);
+      }
+    }
+  }
+
+  drawCropBox(box, context) {
```

Comment:
I don't think this needs to happen on the canvas at all, it can be an element on top of the canvas. This makes it easier to style and control.